File: go-knowledge-app/internal/collectors/hook_inject.go
package collectors

import (
	"bytes"
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sync"
	"time"

	"go-knowledge-app/internal/md"
)

const (
	hookTTLMs     int64 = 60_000
	hookTimeoutMs       = 5_000
)

// HookInjectResult reports the SessionStart hook's live token cost.
type HookInjectResult struct {
	Ok     bool   `json:"ok"`
	Tokens int    `json:"tokens"`
	Output string `json:"output"`
	Error  string `json:"error,omitempty"`
}

// hookRun is one runner invocation's raw result.
type hookRun struct {
	ok     bool
	output string
	err    string
}

// HookRunner spawns the hook; tests pass a fake.
type HookRunner func(cwd string) hookRun

func hookPath() string {
	return filepath.Join(filepath.Dir(ScriptsDir()), "hooks", "inject-project-context.sh")
}

// realRunner spawns `bash <hook>` in cwd with the log override + timeout.
func realRunner(cwd string) hookRun {
	path := hookPath()
	if _, err := os.Stat(path); err != nil {
		return hookRun{ok: false, output: "", err: fmt.Sprintf("Hook nicht gefunden: %s", path)}
	}
	ctx, cancel := context.WithTimeout(context.Background(), hookTimeoutMs*time.Millisecond)
	defer cancel()
	cmd := exec.CommandContext(ctx, "bash", path)
	cmd.Dir = cwd
	// Override the hook's log target so the canonical session log is untouched.
	cmd.Env = append(os.Environ(), "CC_HOOK_LOG_FILE=/dev/null")
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		return hookRun{ok: false, output: "", err: err.Error()}
	}
	return hookRun{ok: true, output: stdout.String() + stderr.String()}
}

type hookCacheEntry struct {
	at    int64
	value HookInjectResult
}

type hookCall struct {
	done  chan struct{}
	value HookInjectResult
}

var (
	hookMu       sync.Mutex
	hookCache    = map[string]hookCacheEntry{}
	hookInflight = map[string]*hookCall{}
)

// CollectHookInject runs the SessionStart hook for cwd, returning {ok, tokens,
// output}. Cached per cwd for hookTTLMs; concurrent calls share one in-flight run.
func CollectHookInject(cwd string) HookInjectResult {
	return collectHookInjectWith(cwd, realRunner)
}

func collectHookInjectWith(cwd string, runner HookRunner) HookInjectResult {
	nowMs := time.Now().UnixMilli()
	hookMu.Lock()
	if e, ok := hookCache[cwd]; ok && nowMs-e.at < hookTTLMs {
		hookMu.Unlock()
		return e.value
	}
	if c, ok := hookInflight[cwd]; ok {
		hookMu.Unlock()
		<-c.done
		return c.value
	}
	c := &hookCall{done: make(chan struct{})}
	hookInflight[cwd] = c
	hookMu.Unlock()

	r := runner(cwd)
	var value HookInjectResult
	if r.ok {
		value = HookInjectResult{Ok: true, Tokens: md.EstTokens(r.output), Output: r.output}
	} else {
		value = HookInjectResult{Ok: false, Tokens: 0, Output: r.output, Error: r.err}
	}

	hookMu.Lock()
	hookCache[cwd] = hookCacheEntry{at: time.Now().UnixMilli(), value: value}
	c.value = value
	delete(hookInflight, cwd)
	close(c.done)
	hookMu.Unlock()
	return value
}

// ResetHookCache drops the per-cwd cache so TTL/single-flight tests start clean.
func ResetHookCache() {
	hookMu.Lock()
	hookCache = map[string]hookCacheEntry{}
	hookInflight = map[string]*hookCall{}
	hookMu.Unlock()
}

File: deno-knowledge-app/src/collectors/hook_inject.ts
// hook_inject — runs the SessionStart hook (inject-project-context.sh) on demand
// and reports its token cost. This is the LIVE part of the Kontext view: it
// captures exactly what the hook would emit into a fresh session in `cwd`.
//
// Safety properties (CCS-031 AC #4):
//   - per-cwd SINGLE-FLIGHT: concurrent callers share one in-flight run
//   - per-cwd TTL cache (60s): repeated requests don't re-spawn the subprocess
//   - 5s TIMEOUT via AbortSignal
//   - CC_HOOK_LOG_FILE=/dev/null so the canonical hook-session-start.log is NOT
//     clobbered by dashboard-triggered runs (the hook honours this override)
//
// The hook output contains only the repo backlog stand + a project tree — NO
// tn (TaskNotes) content and nothing cross-project (Org rule).

import { estTokens } from "../md.ts";

const TTL_MS = 60_000;
const TIMEOUT_MS = 5_000;

export interface HookInjectResult {
  ok: boolean;
  tokens: number;
  output: string;
  /** present only on failure (hook not found / spawn error) */
  error?: string;
}

/** Injectable hook runner — default spawns bash; tests pass a fake. */
export type HookRunner = (cwd: string) => Promise<{ ok: boolean; output: string; error?: string }>;

/** Resolve the hook path relative to this module: src/collectors/ → repo root. */
function hookPath(): string {
  // src/collectors/hook_inject.ts → ../../../hooks/inject-project-context.sh
  return new URL("../../../hooks/inject-project-context.sh", import.meta.url).pathname;
}

/** Real runner: spawn `bash <hook>` in cwd with the log override + timeout. */
const realRunner: HookRunner = async (cwd) => {
  const path = hookPath();
  try {
    await Deno.stat(path);
  } catch {
    return { ok: false, output: "", error: `Hook nicht gefunden: ${path}` };
  }
  try {
    const cmd = new Deno.Command("bash", {
      args: [path],
      cwd,
      // Override the hook's log target so the canonical session log is untouched.
      env: { ...Deno.env.toObject(), CC_HOOK_LOG_FILE: "/dev/null" },
      stdout: "piped",
      stderr: "piped",
      signal: AbortSignal.timeout(TIMEOUT_MS),
    });
    const { stdout, stderr } = await cmd.output();
    const dec = new TextDecoder();
    const output = dec.decode(stdout) + dec.decode(stderr);
    return { ok: true, output };
  } catch (exc) {
    return { ok: false, output: "", error: String(exc) };
  }
};

interface CacheEntry {
  at: number;
  value: HookInjectResult;
}

const cache = new Map<string, CacheEntry>();
const inflight = new Map<string, Promise<HookInjectResult>>();

/**
 * Run the SessionStart hook for `cwd`, returning {ok, tokens, output}.
 * Cached per cwd for {@link TTL_MS}; concurrent calls share one in-flight run.
 * Pass a `runner` to bypass the real subprocess (tests).
 */
export function collectHookInject(
  cwd: string,
  runner: HookRunner = realRunner,
): Promise<HookInjectResult> {
  const now = Date.now();
  const cached = cache.get(cwd);
  if (cached && (now - cached.at) < TTL_MS) {
    return Promise.resolve(cached.value);
  }
  const existing = inflight.get(cwd);
  if (existing) return existing;

  const run = (async () => {
    const r = await runner(cwd);
    const value: HookInjectResult = r.ok
      ? { ok: true, tokens: estTokens(r.output), output: r.output }
      : { ok: false, tokens: 0, output: r.output ?? "", error: r.error };
    cache.set(cwd, { at: Date.now(), value });
    return value;
  })().finally(() => {
    inflight.delete(cwd);
  });
  inflight.set(cwd, run);
  return run;
}

/** Test helper: drop the per-cwd cache so TTL/single-flight tests start clean. */
export function _resetHookCache(): void {
  cache.clear();
  inflight.clear();
}
